pallets/unified-accounts/src/lib.rs

```diff
@@ -0,0 +1,307 @@
+
+    #[pallet::config]
+    pub trait Config: frame_system::Config {
+        /// The overarching event type
+        type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+        /// The Currency for managing evm address assets
+        type Currency: Mutate<Self::AccountId>;
+        /// Default evm address to account id conversion
+        type DefaultAddressMapping: AddressMapping<Self::AccountId>;
+        /// Default account id to evm address conversion
+        type DefaultAccountMapping: AccountMapping<Self::AccountId>;
+        /// The Signature verification implementation to use for checking claims
+        /// Note: the signature type defined by this will be used as parameter in pallet's extrinsic
+        type SignatureHelper: SignatureHelper<AccountId = Self::AccountId, Address = EvmAddress>;
+
+        // /// Weight information for the extrinsics in this module
+        // type WeightInfo: WeightInfo;
+    }
+
+    #[pallet::error]
+    pub enum Error<T> {
+        /// AccountId or EvmAddress already mapped
+        AlreadyMapped,
+        /// The signature is malformed
+        BadSignature,
+        /// The signature verification failed due to mismatch evm address
+        InvalidSignature,
+    }
+
+    #[pallet::event]
+    #[pallet::generate_deposit(pub(super) fn deposit_event)]
+    pub enum Event<T: Config> {
+        /// Evm Address claimed.
+        /// Double Mapping b/w native and evm address created
+        AccountClaimed {
+            account_id: T::AccountId,
+            evm_address: EvmAddress,
+        },
+    }
+
+    /// Native accounts for evm address
+    /// NativeToEvm: EvmAddress => Option<AccountId>
+    #[pallet::storage]
+    pub type NativeToEvm<T: Config> =
+        StorageMap<_, Blake2_128Concat, EvmAddress, T::AccountId, OptionQuery>;
+
+    /// evm addresses for native accounts
+    /// EvmToNative: AccountId => Option<EvmAddress>
+    #[pallet::storage]
+    pub type EvmToNative<T: Config> =
+        StorageMap<_, Blake2_128Concat, T::AccountId, EvmAddress, OptionQuery>;
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        /// Claim account mapping between Substrate account and Evm address.
+        /// Ensure no prior mapping exists for evm address.
+        ///
+        /// - `evm_address`: The evm address to bind to the caller's account
+        /// - `signature`: A signature generated by the address to prove ownership
+        #[pallet::call_index(0)]
+        #[pallet::weight(0)]
+        pub fn claim_evm_address(
+            origin: OriginFor<T>,
+            evm_address: EvmAddress,
+            signature: SignatureOf<T>,
+        ) -> DispatchResult {
+            let who = ensure_signed(origin)?;
+            // make sure no prior mapping exists
+            Self::enure_no_mapping(&who, &Some(evm_address))?;
+            // claim the evm address
+            Self::do_claim_evm_address(who, evm_address, signature)
```

Comment:
As this is basically copy & paste from Acala, pls keep it the same as the original code and DO NOT restructure it.

The Acala impl was [audited by SRLabs](https://github.com/AcalaNetwork/Acala/blob/master/audit/SRL_Acala-review-2021-12-public.pdf) and has been battle-tested for some time. We don't want to introduce easily overlooked vulnerability by restructuring and adding minor tweaks.

Please also double check unit tests and make sure they covered existing tests in Acala.

Reply:
Sure, I'll try to match Acala implementation

> Please also double check unit tests and make sure they covered existing tests in Acala.

All the units are covered with more tests for individual implementations as well.